### PR TITLE
minstall: allow missing symlinks in `install_symlink`

### PR DESCRIPTION
**Pull request by Hi-Angel**
Currently there's an undocumented behavior where using `install_symlink()` to point to a file that does not exist results in failure `ERROR: Tried to install symlink to missing file`.

Such behavior makes little sense, because:

1. There's no reason the destination file should exist. The "destination file" may belong to another package, in particular to a package inside the same repo the Meson is controlling, but in both cases there's no reason the other package should be present in the system, because installation does not typically happen into the system.
2. Since installation happens into DESTDIR and not into the system (barring the fact the DESTDIR may be the / which is frowned upon anyway), the code conceptually should rather check for "destination file" presence in the DESTDIR, not in the system, so even if we do want to check for file presence, it is currently done incorrectly.

In any case, best backward-compatible decision here would be to just remove the check, which is done by this commit.

Fixes: https://github.com/mesonbuild/meson/issues/12253

### CODE REVIEW

**Comment by Hi-Angel** [comment]
Hmm, well, I need to reformulate the point `2`, because it does check the DESTDIR for destination presence. I was thinking of something else: when you have 2 or more packages inside single Meson repo, you end up having different `DESTDIR`s *(one for each package)*, and of course Meson wouldn't check `DESTDIR`s for other packages. It wouldn't be reliable anyway, because what if the other package wasn't "ninja-install"ed yet.

Okay, let me rephrase that part.

**Comment by Hi-Angel** [comment]
Done. I just merged both points in explanation.

**Comment by Hi-Angel** [comment]
Hello, any news here?

**Comment by eli-schwartz** [comment]
I'm not entirely sure I understand the general use case for installing a broken symlink. Am I missing something here?

**Comment by Hi-Angel** [comment]
Well, in the project where I'm trying to use this Meson functional *(we currently use a custom script instead)* I see 3 different packages where we create symlinks, in one of them symlink is being created to a file that is apparently autogenerated during runtime, and another is to point to a file that belongs to a different package. So you can say, 66% of symlinks in our project are pointing to a file that will only appear after a package is installed, not during the meson-running stage.

**Comment by Hi-Angel** [comment]
…but as far as "the general use case" goes, I don't have much to add besides what I mentioned in the top post. It seems pretty usual to me that a package you want to create a symlink to may just not be present in the system, or alternatively it is a package that is part of the Meson-managed repo you're in, but one that belongs to a different DESTDIR. Do you think there's something strange with that?

**Comment by nils-werner** [comment]
I'm not sure if I'm misunderstanding how symlinks are treated during isntallation, but doesn't basically every symlink with an absolute path result in a dead link during build time? Don't you always have to do

```
ln -s \
    /usr/lib/systemd/system/my-service.service \
    $DESTDIR/etc/systemd/system/multi-user.target.wants/my-service.service
```

i.e. the link is inside `$DESTDIR` but points outside of it.

**Comment by Hi-Angel** [comment]
@nils-werner I am not sure I understood your question, but symlink with an absolute path may or may not be a deadlink. You provide example with absolute paths, so in your example depending on whether `/usr/lib/systemd/system/my-service.service` existed or not it would be "dead link" or not.

**Comment by nils-werner** [comment]
@Hi-Angel exactly, the fact that an absolute link is valid is an unintended side-effect of the build environment, not an intended effect of the target environment. Right now it seems Meson doesn't allow the creation of absolute links, unless it isn't dead because of this unintended side-effect.

**Comment by eli-schwartz** [comment]
This ticket is about symlinks that are installed to $DESTDIR and then a package manager tarballs up $DESTDIR and records the files and then extracts the files to a root filesystem and then, on the final root filesystem, the symlink is dead because the package only installs the symlink but doesn't install the destination.

**Comment by nils-werner** [comment]
@eli-schwartz are you sure? I understand it the other way round: the symlink is dead during `meson install`, but later on the target system it isn't.

**Comment by Dudemanguy** [comment]
Any updates on this one? I unexpectedly ran into the exact limitation here with the newest version of openrc because it tries to install symlinks that may not exist on the system. With this patch, it works fine again. I don't understand why this check exists?

**Comment by bonzini** [comment]
I would prefer to add `allow_missing` as a new keyword argument to the function. It's a safeguard against bugs.

**Comment by eli-schwartz** [comment]
It has been like this since the initial implementation in #9608

As for openrc:

```
Installing symlink pointing to /etc/init.d/agetty to /home/eschwartz/git/openrc/builddir/pkg/etc/init.d/agetty.tty1

ERROR: Tried to install symlink to missing file /home/eschwartz/git/openrc/builddir/pkg/etc/init.d/agetty.tty1
```

and:
```
$ ls -l builddir/pkg/etc/init.d/
total 164
-rwxr-xr-x 1 eschwartz eschwartz 1242 Apr 25 12:10 agetty*
lrwxrwxrwx 1 eschwartz eschwartz   18 Apr 25 12:11 agetty.tty1 -> /etc/init.d/agetty
```

Latter case is blinking red, because it doesn't exist yet. We are trying to create absolute symlinks in a DESTDIR where one points to the other, so in fact this is expected to not exist at staged-install time (DESTDIR) but will exist at real-install time (merging DESTDIR to the live filesystem e.g. via a distro package manager).

So yeah, this check makes no sense I guess.

**Comment by bonzini** [comment]
> we need a more sophisticated check that handles recursively checking broken symbolic links against their DESTDIR offset.

Yes, I agree that this is broken for DESTDIR so it can be treated as a bugfix. Adding to the 1.9 milestone.

**Comment by eli-schwartz** [comment]
(The other question is whether these should be absolute or relative paths. Relative ones won't be broken symlinks at install time, so the verification check would pass anyway.)

**Comment by Hi-Angel** [comment]
rebased, CI seems fully passed.
